### DISCUSSION TITLE
InvalidDatasourceError: Datasource URL must use prisma:// protocol

### DISCUSSION THREAD

**Discussion by Danubius1st** [prisma/prisma#20240] [Q&A] [answered]
I develop an Nextjs app. The app is working as espected locally, both in development and production mode, using data in cloud. The build on vercel works, the app works, but if I try to use my database I get this error on logs:
`InvalidDatasourceError: Datasource URL must use prisma:// protocol when --accelerate or --data-proxy are used
`
and
`api/graphql: Failed to load resource: the server responded with a status of 405`
on browser console.

/graphql/builder.ts:
```
import SchemaBuilder from '@pothos/core';
import PrismaPlugin from '@pothos/plugin-prisma';
import RelayPlugin from '@pothos/plugin-relay';
import type PrismaTypes from '@pothos/plugin-prisma/generated';
import prisma from '../lib/prisma';

export const builder = new SchemaBuilder<{
  PrismaTypes: PrismaTypes;
}>({
  plugins: [PrismaPlugin, RelayPlugin],
  relayOptions: {},
  prisma: {
    client: prisma,
  },
});

builder.queryType({});
builder.mutationType({});

```
/lib/apollo.ts:

```
import { ApolloClient, InMemoryCache } from '@apollo/client';

export const apolloClient = new ApolloClient({
  uri: '/api/graphql',
  cache: new InMemoryCache(),
});

```
/lib/prisma.ts:

```
import { PrismaClient } from '@prisma/client';

let prisma: PrismaClient;

declare global {
  var prisma: PrismaClient;
}

if (process.env.NODE_ENV === 'production') {
  prisma = new PrismaClient();
} else {
  if (!global.prisma) {
    global.prisma = new PrismaClient();
  }
  prisma = global.prisma;
}

export default prisma;

```
/api/graphql.ts:

```
import { schema } from '../../graphql/schema';
import { createYoga } from 'graphql-yoga';
import { NextApiRequest, NextApiResponse } from 'next';

export default createYoga<{
  req: NextApiRequest;
  res: NextApiResponse;
}>({
  schema,
  graphqlEndpoint: process.env.GRAPHQL_API,
});

export const config = {
  api: {
    bodyParser: false,
  },
};

```
/prisma/schema.prisma:

```
generator client {
  provider = "prisma-client-js"
}

generator pothos {
  provider = "prisma-pothos-types"
}

datasource db {
  provider     = "mysql"
  url          = env("DATABASE_URL")
  relationMode = "prisma"
}
...

```

.env:

`DATABASE_URL=mysql://<user>:<pass>@<host>/<myDB>?sslaccept=strict
`
If I change mysql:// to prisma:// the result is the same.

package.json:

```
...
next@13.4.10
graphql@16.7.1
graphql-yoga@4.0.3
@apollo/client@3.7.17
@apollo/server@4.7.5
@prisma/client@5.0.0
prisma@5.0.0
...
```
And production build script:

`    "vercel-build": "npx prisma generate --data-proxy && next build",`

 I also noticed that planetscale changed the host from us-east.connect.psdb.cloud to aws.connect.psdb.cloud. I don't know if this has to do with the regions where Data Proxy infrastructure is currently available.

I don't know if this is a bug, but I've run out of solutions. Please help.

**Comment by janpio** [comment]
What value is `DATABASE_URL` set to in your Vercel project? `mysql://...` or `prisma://...`?

**Comment by Danubius1st** [comment]
`mysql://...`

**Reply by nurul3101** [reply, marked as answer]
@Danubius1st I have responded on your Stack Overflow question:
https://stackoverflow.com/questions/76690384/invaliddatasourceerror-datasource-url-must-use-prisma-protocol

Just duplicating here as well:

You don't need to replace `mysql://` with `prisma://`. Have you signed up for accelerate from this [URL](https://cloudprojects.prisma.io/login)? If not, you would need to sign up and create a new project. The project will ask you to submit your database URL, you would need to provide the one starting with mysql://. After which you would get a new connection string that would start with `prisma:/`/ that you need to use.

In general, when using Data Proxy (or Accelerate), you'll be dealing with two different connection strings:

- **Database connection string**: This one is the connection string that enables direct access to your DB and starts with `mysql://...`. It is used to run migrations and you can add it to your `datasource` using the [`directUrl`](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#external-connection-poolers) field.
- **Data Proxy (or Accelerate) connection string**: This is the connection string that connects to the connection pooler and needs to be specified as the `url` of your `datasource` block as it's used by Prisma Client to send queries to the DB. It typically looks as follows: `prisma://accelerate.prisma-data.net/?api_key=__API_KEY__`

**Reply by nurul3101** [reply]
@Danubius1st Let us know in case you are still running into this issue.

It would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 
Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by Danubius1st** [reply]
See [comment](https://stackoverflow.com/questions/76690384/invaliddatasourceerror-datasource-url-must-use-prisma-protocol)

**Reply by nurul3101** [reply]
I am glad to hear that you got it working.

I'll mark this discussion as resolved.
